### ISSUE TITLE
Suggest problems for people to port to another language

### ISSUE THREAD

**Issue by kytrinyx** [exercism/rossconf#15]
Goal: https://github.com/exercism/rossconf/issues/9

-------------

**Skillsets:** A fairly solid working knowledge of Ruby or JavaScript/Angular. This would probably be a fairly localized change inside the exercism.io application... some code to talk to an API, some view code, maybe some presenter code.

-------------

There are two strategies that I can think of that might be useful:

1. When someone has completed several (all?) exercises on a given language track, we can suggest that they port some more exercises to that language.
2. When someone has completed exercises in more than one language, find exercises that they've done in one language that don't yet exist in another language that they have done, and suggest that they contribute that exercise.

This talk would consume the APIs exposed in #14 in order to make suggestions to the user (perhaps on their dashboard).

**Comment by rubydog** [comment]
Hey kytrinyx, I would like to take a look if I can implement this feature, can you suggest where can I start? and help me setup project on my machine. Thanks!

**Comment by kytrinyx** [comment]
To set up the project on your machine, you'll need to start with the instructions here: https://github.com/exercism/exercism.io/blob/master/CONTRIBUTING.md#setup

When you say _suggest where you can start_ do you mean how to tackle the feature? Or where to find the relevant parts of the code?

**Comment by rubydog** [comment]
Both I guess, I have already setup the project.

**Comment by kytrinyx** [comment]
This is a really big feature. I would suggest either picking a smaller feature to start with so that you get familiar with the codebase, or maybe you can define a small "first version" of this feature.

**Comment by GustavoCaso** [comment]
@kytrinyx I would to work on this one.

Do we have any design or roadmap that we what to achieve ?

**Comment by kytrinyx** [comment]
Hey @GustavoCaso -- this is definitely right up your alley! We don't have a design yet, but I'm happy to throw some ideas back and forth with you here, if you want to kick it off.

**Comment by GustavoCaso** [comment]
@kytrinyx So what do you think if this show in the profile page, to an use if the requirements are met, ex: If the user has finish all the exercise of a track, it could see a link to the new contribute page or if the user has multiple exercise from different tracks, and suggest a non implemented problem, right there ?

**Comment by kytrinyx** [comment]
Good question. I don't know if the profile page or the dashboard would be a better place for it. Profile page seems like a good idea, since this is a more "long term" bit of data.

I think both of your ideas are good, and could be implemented separately (one PR to do link to the contribute page if they've finished all the exercises in a track, and a separate one for suggesting a non-implemented problem).

**Comment by GustavoCaso** [comment]
Great @kytrinyx I will work on both PR

**Comment by GustavoCaso** [comment]
@kytrinyx I have been thinking about this type of specific contributions:

> If someone has done the same exercise in more than one language, suggest that they compare the assertions in both/all of the languages, and if there are differences, see if there are improvements that can be made. They could even help extract common inputs/outputs to a json file in our metadata repository

And I think we can query the database to extract that information, but after getting the information I don't know what we could show the user, any idea?

Also with this one: 
> If someone has done exercises in more than one language, find problems that they've solved in one of their languages, that doesn't exist in another of their languages. Suggest that they port the exercise.

I'm thinking about using the `todo` endpoint to randomly check for a track where the problem is in the `todo` list. So I thought maybe we could open an endpoint to look for `todos` instead by `track` by `slug` name. What do you think ? I'll be really happy to add the PR to the `x-api`

**Comment by GustavoCaso** [comment]
☝️  Never mind with the last part, I just show that we already have that endpoint implemented, so I will use that one to show random messages.

**Comment by kytrinyx** [comment]
Excellent. Yeah, we have an endpoint that has the data.

I think the thing we can show them is something like this:

- detect that they've solved problem X in language A (and maybe B, C, D).
- detect that they've solved some problems in a language E that doesn't have problem X.

For example:

> You've solved **Clock** in Ruby and Go. Elixir doesn't have **Clock** yet. Want to implement it?

And then _implement it_ would link to the instructions for how to port an exercise to another language.

We would only show one at a time.

This is likely to be a bit time-consuming, so we might want to create a separate table (or use redis) and have a job that finds all the possible matches for you, and then just pick one of them to show randomly each time we show the dashboard.

**Comment by GustavoCaso** [comment]
Perfect I will have a look ASAP. And create a PR for each kind of
suggestions. 😃

**Comment by kytrinyx** [comment]
I'm not sure what you mean by _each kind of suggestions_ - do you have several different ideas about how to solve this?

**Comment by GustavoCaso** [comment]
@kytrinyx maybe I didn't explain myself very well, I meant create a PR for the suggestions with the spec:
> If someone has done exercises in more than one language, find problems that they've solved in one of their languages, that doesn't exist in another of their languages. Suggest that they port the exercise.

And another PR for the suggestions with this specs:
> If someone has done the same exercise in more than one language, suggest that they compare the assertions in both/all of the languages, and if there are differences, see if there are improvements that can be made. They could even help extract common inputs/outputs to a json file in our metadata repository

**Comment by kytrinyx** [comment]
OK, thank you. That sounds good!

**Comment by GustavoCaso** [comment]
@kytrinyx sorry for the long wait, I've working a lot lately, but I going to continue with this ones, I have some doubt about something that you mentioned before.
> This is likely to be a bit time-consuming, so we might want to create a separate table (or use redis) and have a job that finds all the possible matches for you

So if you wanted to create a separate table, we will have to store that information every time a user finish an exercises  right ? So we can query this table to extract for example the slug of a problem that the user has completed in multiple languages and check if a track is missing this one.

That sound good to you ?

**Comment by kytrinyx** [comment]
Yeah, we will have to store the information every time, but this is an add-on. If we don't have data for someone yet, we won't show them anything. If we do, then we won't need a spinner.
